### ISSUE TITLE
課題2-2 追加テーブルへのデータ保存

### ISSUE THREAD

**Issue by chika-O** [chika-O/lamp_practice#12]
purchase_details/purchase_historiesへのデータ保存部分を実装します。

**Comment by chika-O** [comment]
@primestyleus 
追加テーブルへのデータ保存部分を以下の手順で実装します。
・model/cart.phpに、テーブルへの保存のため関数regist_purchase_informationを記述します。ここでトランザクションを用いて二つのテーブルにデータを保存する処理を記述します。
・finish.phpでregist_purchase_informationを呼び出し、$cartsを用いて取得したデータを登録します。

以上です。
ご確認をお願いいたします。

**Comment by primestyleus** [comment]
@chika-O 
はい、こちらでOKですが、トランザクションまで考慮すると、少し複雑になるかもしれません。model/cart.phpにある既存の関数に追加する形だと、もう少しシンプルにできますが、まずは実装を進めてみてください

**Comment by chika-O** [comment]
@primestyleus 
お世話になっております。課題2-2についての質問です。
以下のようにmodel/cart.phpに追記をしたのですが、purchase_cartsにregist_purchase_histories/regist_purchase_detailsを追記すると、delete_user_cartsも含めてexecute_queryが実行されず、全て例外処理で返ってくるようになってしまいました。
どこに原因があると考えられますでしょうか？

`// 購入処理
function purchase_carts($db, $carts){
  // 正常に購入できる状態かどうかをチェック
  if(validate_cart_purchase($carts) === false){
    return false;
  }
  $db->beginTransaction();


  // amount分を在庫から削除する
  foreach($carts as $cart){
    if(update_item_stock(
        $db, 
        $cart['item_id'], 
        $cart['stock'] - $cart['amount']
      ) === false){
      set_error($cart['name'] . 'の購入に失敗しました。');
    }
  }
  // 該当ユーザのカート情報を削除
  delete_user_carts($db, $carts[0]['user_id']);

  //追加テーブルへのデータ書き込み
  regist_purchase_histories($db,$carts);
  regist_purchase_details($db,$carts);
  

  //set_errorされたかをを判別
  if(has_error() === false){
    $db->commit();
  }else {
    $db->rollback();
  }
}

/purchase_historiesの更新
function regist_purchase_histories($db,$carts){
    $sql = "
      INSERT INTO
        purchase_histories(
          user_id
        )VALUES(
          :user_id
        )
    ";
    $params = array(':user_id' => $user_id);
    execute_query($db, $sql, $params);

  }

  //purchase_detailsの更新
  function regist_purchase_details($db,$carts){
    foreach ($carts as $cart) {
      $sql = "
        INSERT INTO
          purchase_details(
            purchase_id,
            item_id,
            purchase_price,
            purchase_amount
          )VALUES(
            :purchase_id,
            :item_id,
            :purchase_price,
            :purchase_amount
          )
      ";
    
      $purchase_id = $db->lastInsertId();
      $params = array(':purchase_id' => $purchase_id,':item_id' => $cart['item_id'],':purchase_price' =>$cart['price'],':purchase_amount' => $cart['amount']);
      execute_query($db, $sql, $params);
    }
  }`

**Comment by primestyleus** [comment]
@chika-O 
そうですね、ぱっとみて気になるのは以下2点です。

■regist_purchase_histories関数
`$params = array(':user_id' => $user_id);`
$user_idはこの関数の中では定義されていないので$cartsの中からうまくとってくる必要があります。

■ regist_purchase_details関数
`$purchase_id = $db->lastInsertId();`
これについてはregist_purchase_historiesを実行したあとに1回だけ実行して、引数でregist_purchase_detailsに渡すとうまくいくと思います。

まずは上記について修正してうまくいくか試してみてください！
